resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/StagedResourcesCleaner.scala

```diff
@@ -0,0 +1,154 @@
+      clock: Clock,
+      initialAccessExpirationMs: Long,
+      resourceCleanupIntervalMs: Long)
+    extends StagedResourcesCleaner {
+
+  private val RESOURCE_LOCK = new Object()
+  private val activeResources = mutable.Map.empty[String, MonitoredResource]
+  private val unusedResources = mutable.Map.empty[String, UnusedMonitoredResource]
+
+  override def start(): Unit = {
+    cleanupExecutorService.scheduleAtFixedRate(
+        new CleanupRunnable(),
+        resourceCleanupIntervalMs,
+        resourceCleanupIntervalMs,
+        TimeUnit.MILLISECONDS)
+  }
+
+  override def registerResourceForCleaning(
+      resourceId: String, stagedResourceOwner: StagedResourcesOwner): Unit = {
+    RESOURCE_LOCK.synchronized {
+      unusedResources(resourceId) = UnusedMonitoredResource(
+          clock.getTimeMillis() + initialAccessExpirationMs,
+          MonitoredResource(resourceId, stagedResourceOwner))
+
+    }
+  }
+
+  override def markResourceAsUsed(resourceId: String): Unit = RESOURCE_LOCK.synchronized {
+    val resource = unusedResources.remove(resourceId)
+    resource.foreach { res =>
+      activeResources(resourceId) = res.resource
+    }
+  }
+
+  private class CleanupRunnable extends Runnable with Logging {
+
+    override def run(): Unit = {
+      // Make a copy so we can iterate through this while modifying
+      val activeResourcesCopy = RESOURCE_LOCK.synchronized {
+        Map.apply(activeResources.toSeq: _*)
+      }
+      for ((resourceId, resource) <- activeResourcesCopy) {
+        Utils.tryWithResource(
+            kubernetesClientProvider.getKubernetesClient(
+                resource.stagedResourceOwner.ownerNamespace,
+                resource.stagedResourceOwner.ownerMonitoringCredentials)) { kubernetesClient =>
+          val namespace = kubernetesClient.namespaces()
+              .withName(resource.stagedResourceOwner.ownerNamespace)
+              .get()
+          if (namespace == null) {
+            logInfo(s"Resource with id $resourceId is being removed. The owner's namespace" +
+              s" ${resource.stagedResourceOwner.ownerNamespace} was not found.")
+            stagedResourcesStore.removeResources(resourceId)
+            RESOURCE_LOCK.synchronized {
+              activeResources.remove(resourceId)
+            }
+          } else {
+            val metadataOperation = resource.stagedResourceOwner.ownerType match {
+              case StagedResourcesOwnerType.Pod =>
+                kubernetesClient.pods()
+              case _ =>
+                throw new SparkException(s"Unsupported resource owner type for cleanup:" +
+                  s" ${resource.stagedResourceOwner.ownerType}")
+            }
+            if (metadataOperation
+              .withLabels(resource.stagedResourceOwner.ownerLabels.asJava)
+              .list()
+              .getItems
+              .isEmpty) {
+              logInfo(s"Resource with id $resourceId is being removed. Owners of the resource" +
```

Comment:
Changed the log wording to indicate we're removing the files and not Kubernetes objects.